packages/expo-screen-orientation/ios/EXScreenOrientation/EXScreenOrientationRegistry.m

```diff
@@ -0,0 +1,227 @@
+    return;
+  }
+  
+  UIInterfaceOrientationMask currentOrientationMask = [self currentOrientationMask];
+  
+  // checks if screen orientation should be changed when user rotates the device
+  if ([EXScreenOrientationUtilities doesOrientationMask:currentOrientationMask containOrientation:newScreenOrientation]) {
+    // change current screen orientation
+    if ((UIInterfaceOrientationIsPortrait(newScreenOrientation) && UIInterfaceOrientationIsPortrait(_currentScreenOrientation))
+        || (UIInterfaceOrientationIsLandscape(newScreenOrientation) && UIInterfaceOrientationIsLandscape(_currentScreenOrientation))) {
+      _currentScreenOrientation = newScreenOrientation; // updates current screen orientation, but doesn't emit event
+      return;
+    }
+    
+    // on iPads, traitCollectionDidChange isn't triggered at all
+    if ([EXScreenOrientationUtilities isIPad] &&
+        ((UIInterfaceOrientationIsPortrait(newScreenOrientation) && UIInterfaceOrientationIsLandscape(_currentScreenOrientation))
+         || (UIInterfaceOrientationIsLandscape(newScreenOrientation) && UIInterfaceOrientationIsPortrait(_currentScreenOrientation)))) {
+      [self screenOrientationDidChange:newScreenOrientation];
+    }
+  }
+}
+
+- (void)traitCollectionsDidChange:(UITraitCollection *)traitCollection
+{
+  UIUserInterfaceSizeClass verticalSizeClass = traitCollection.verticalSizeClass;
+  UIUserInterfaceSizeClass horizontalSizeClass = traitCollection.horizontalSizeClass;
+  UIInterfaceOrientation currentDeviceOrientation = [EXScreenOrientationUtilities UIDeviceOrientationToUIInterfaceOrientation:[[UIDevice currentDevice]  orientation]];
+  UIInterfaceOrientationMask currentOrientationMask = [[[[UIApplication sharedApplication] keyWindow] rootViewController] supportedInterfaceOrientations];
+   
+  UIInterfaceOrientation newScreenOrientation = UIInterfaceOrientationUnknown;
+
+  if (verticalSizeClass == UIUserInterfaceSizeClassRegular && horizontalSizeClass == UIUserInterfaceSizeClassCompact) {
+    // portrait
+    UIInterfaceOrientationMask portraitMask = currentOrientationMask & UIInterfaceOrientationPortrait;
+    if (portraitMask == UIInterfaceOrientationMaskPortrait) {
+      newScreenOrientation = UIInterfaceOrientationPortrait;
+    } else if (portraitMask == UIInterfaceOrientationMaskPortraitUpsideDown) {
+      newScreenOrientation = UIInterfaceOrientationPortraitUpsideDown;
+    } else if (currentDeviceOrientation == UIInterfaceOrientationPortrait || currentDeviceOrientation == UIInterfaceOrientationPortraitUpsideDown) {
+      newScreenOrientation = currentDeviceOrientation;
+    }
+  } else if ((verticalSizeClass == UIUserInterfaceSizeClassCompact && horizontalSizeClass == UIUserInterfaceSizeClassCompact)
+              || (verticalSizeClass == UIUserInterfaceSizeClassCompact && horizontalSizeClass == UIUserInterfaceSizeClassRegular)){
```

Comment:
If we include case of (compact == compact) should we also handle case (regular == regular), just in case?